### PR TITLE
[WIP - Rebranch] Collab/cwp calib tutorial

### PR DESCRIPTION
**Pull request by jlnav**
Current state:

https://libensemble.readthedocs.io/en/collab-cwp_calib_tutorial_rebranch/tutorials/cwp_calib_tutorial.html

This is a new branch containing _only_ tutorial content, instead of content also on Develop.

If it's alright, for now I'll keep the other branch open for reference, since there's lots of good suggestions over there. I'll then close it without merging once the Tutorial on this branch is ready.

### CODE REVIEW

File: docs/tutorials/cwp_calib_tutorial.rst [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+
+Introduction - Calibration with libEnsemble and CWP
+---------------------------------------------------
+
+This tutorial demonstrates libEnsemble's capability to selectively cancel pending
+simulations based on instructions from the *Persistent CWP* calibration
+generator function. This capability is critical for this calibration use-case since
+it isn't useful for the generator to receive extraneous evaluations
+from resources that may be more effectively applied towards critical evaluations.
+
```
Comment (shuds13) [inline docs/tutorials/cwp_calib_tutorial.rst:13]: I wouldn't say "critical", but improves efficiency. Maybe simplfy the sentence a bit.

e.g. This improves the efficiency by preventing time spent running obviated simulations.

File: docs/tutorials/cwp_calib_tutorial.rst [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+For a somewhat different approach than libEnsemble's :doc:`other tutorials<tutorials>`,
+we'll emphasize the settings, functions, and data fields within the calling script, CWP
+:ref:`persistent generator<persistent-gens>`, Manager, and :ref:`sim_f<api_sim_f>`
+that make this capability possible, rather than outlining a step-by-step process
+for writing this exact use-case.
+
+.. note::
+    The generator function featured in this tutorial can be found in ``gen_funcs/persistent_cwp_calib.py``. This version uses simplified standalone routines in place of the in-development CWP library.
+
```
Comment (shuds13) [inline docs/tutorials/cwp_calib_tutorial.rst:24]: Wondering if maybe some psuedo-code of gen would help???? See what you think @jlnav @mosesyhc 
Something like

* Calculate test values and first batch
* While no STOP signal received:
  * Unpack evaluated points into 2D theta x point structures
  * If new model condition:
    * Construct new model
  * else
    * wait to receive more points
  * if some condition:
    * Generate new points from model
    * If threshold error is reached
      * Exit loop - done
  * If issued points are obviated
    * Cancel points

Could even show data layout, not sure if needed?

H:
 theta1, x1
 theta1, x2
 theta1, x3
 ...
 
->

fevals
 Row(theta 1):  x1, x2, x3, x4, x5
 Row(theta 2):  x1, x2, x3, x4, x5
 Row(theta 3):  x1, x2, x3, x4, x5
 ...

failures, and data_status same layout.

File: docs/tutorials/cwp_calib_tutorial.rst [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+While the generator loops and updates the model based on returned
+points from simulations, it detects using a library function if any pending points
+and Thetas distributed for simulation are no longer needed for the model
+and ought to be cancelled (obviated). The generator then calls ``cancel_row()``::
+
+    r_obviate = obviate_pend_thetas(model, theta, data_status)
+    if r_obviate[0].shape[0] > 0:
+        cancel_row(pre_count, r_obviate, n_x, data_status, comm)
+
+``pre_count`` is a matrix of Thetas and "x"s, ``r_obviate`` is a selection
```
Comment (shuds13) [inline docs/tutorials/cwp_calib_tutorial.rst:47]: pre_count is a scalar of the number of sim_ids to used for the test values, basically an offset.

File: docs/tutorials/cwp_calib_tutorial.rst [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+
+Within ``cancel_row()``, each row in ``r_obviate`` is iterated over, and if a
+point's ``data_status`` indicates it has not yet been evaluated by a simulation,
+it's ``sim_id`` is appended to a list to be sent to the Manager for cancellation.
+A new, separate local :doc:`History array<../history_output>` is defined with the
+selected ``'sim_id'`` s and the ``'cancel_requested'`` field set to ``True``. This array is
+then sent to the Manager using the ``send_mgr_worker_msg`` persistent generator
+helper function. Each of these helper functions is described :ref:`here<p_gen_routines>`.
+The entire ``cancel_row()`` routine is listed below::
+
```
Comment (shuds13) [inline docs/tutorials/cwp_calib_tutorial.rst:60]: I've updated this func to cancel_rows now.

File: docs/tutorials/cwp_calib_tutorial.rst [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+        send_mgr_worker_msg(comm, H_o)
+
+Most Workers, including those running other persistent generators, are only
+allocated work when they're in an :doc:`idle or non-active state<../data_structures/worker_array>`.
+However, since this generator must asynchronously update its model and
+cancel pending evaluations, the Worker running this generator remains
+in a unique *active receive* state, until it becomes non-persistent. This means
+both the Manager and persistent Worker must be prepared for irregular sending /
+receiving of data.
+
```
Comment (shuds13) [inline docs/tutorials/cwp_calib_tutorial.rst:86]: I would probably not say unique, as it wont always be.

File: docs/tutorials/cwp_calib_tutorial.rst (hunk 1 above)
Comment (jlnav) [reply docs/tutorials/cwp_calib_tutorial.rst:13]: Thank you, I mention the irregular comms patterns in the end of the Generator section, but I could add an introductory remark near the beginning, and expand more later on.

Comment (shuds13) [reply docs/tutorials/cwp_calib_tutorial.rst:13]: Maybe just one sentence to mention uses irregular comms here

**Comment by shuds13** [comment]
I added a few thoughts/fixes. Overall, I think this is coming on well.

**Comment by mosesyhc** [comment]
I put some clarifications about calibration in general, and delineating the problem itself and the specific way cwp tries to implement. The ability for user to swap out those functionalities with their own seems important to note. 

There are some other cwp references, e.g. code file namings, to clean up. I would probably do that when the model file is up.

**Pull request closed without merging**